server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesFetcher.java

```diff
@@ -100,7 +103,19 @@ private FieldCapabilitiesIndexResponse doFetch(
         }
 
         final MappingMetadata mapping = indexService.getMetadata().mapping();
-        final String indexMappingHash = mapping != null ? mapping.getSha256() : null;
+        String indexMappingHash;
+        if (includeFieldsWithNoValue) {
+            indexMappingHash = mapping != null ? mapping.getSha256() : null;
+        } else {
+            // even if the mapping is the same if we return only fields with values we need
+            // to make sure that we consider all the shard-mappings pair, that is why we
+            // calculate a different hash for this particular case.
+            StringBuilder sb = new StringBuilder(indexService.getShard(shardId.getId()).getShardUuid());
+            if (mapping != null) {
+                sb.append(mapping.getSha256());
+            }
+            indexMappingHash = sb.toString();
```

Comment:
I'll test it out and get back to you, it seems that `we force the other cluster to not merge the results but instead send us back all individual index results` therefore this change will duplicate mappings. 
I am thinking we could let the cluster merge the results in case this flag is active. 
I'll create a test in `CCSFieldCapabilitiesIT` and keep you posted.

Reply:
This do not affect CCS and de-duplicating is still working fine. 
Normal de-duplication is done through hash first and by index name later (`indexResponses.putIfAbsent(resp.getIndexName(), resp)`). 
In case the flag is set and we have multiple responses we perform a merge of just the field capabilities, returning one response per index with mapping and field capabilities merged from all intra-cluster index responses.

Reply:
This assessment does not sound accurate to me. The mapping hash is used to deduplicate mappings, and decide to send back one mapping per hash instead one mapping per index. But the whole mapping needs to be sent back?

Using the flag, the index mapping hash deduplication is entirely disabled. We do have now field level deduplication across indices though, and the benefit that not returning empty fields will make responses smaller and decrease overhead on the merging side. But again, it is an entirely different trade-off and we need to measure the impact in the worst case scenario, compared to the previous index mapping hash based deduplication. Makes sense?